python/packages/ai/teams/app.py

```diff
@@ -647,76 +647,93 @@ async def on_turn(self, context: TurnContext):
+                return
 
-            await state.save(context, self._options.storage)
+            await self._save_state(context, state)
         except ApplicationError as err:
             await self._on_error(context, err)
         finally:
             self.typing.stop()
 
+    async def _start_typing(self, context: TurnContext):
+        if self._options.start_typing_timer:
+            await self.typing.start(context)
+
+    async def _remove_mentions(self, context: TurnContext):
+        if self.options.remove_recipient_mention and context.activity.type == ActivityTypes.message:
+            context.activity.text = context.remove_recipient_mention(context.activity)
+
+    async def _initialize_state(self, context: TurnContext):
+        if self._turn_state_factory:
+            state = await self._turn_state_factory(context)
+        else:
+            state = cast(StateT, await TurnState.load(context, self._options.storage))
+        await state.load(context, self._options.storage)
+        state.temp.input = context.activity.text
+        return state
+
+    async def _run_before_turn_middleware(self, context: TurnContext, state):
+        for before_turn in self._before_turn:
+            is_ok = await before_turn(context, state)
+            if not is_ok:
+                await self._save_state(context, state)
+                return False
+        return True
+
+    async def _handle_file_downloads(self, context: TurnContext, state):
+        if self._options.file_downloaders and len(self._options.file_downloaders) > 0:
+            input_files = state.temp.input_files if state.temp.input_files is not None else []
+            for file_downloader in self._options.file_downloaders:
+                files = await file_downloader.download_files(context)
+                input_files.append(files)
+            state.temp.input_files = input_files
+
+    async def _run_activity_handlers(self, context: TurnContext, state):
+        is_ok, matches = await self._on_activity(context, state)
+        if not is_ok:
+            await self._save_state(context, state)
+        return is_ok, matches
+
+    async def _run_ai_chain(self, context: TurnContext, state):
+        if (
+            self._ai
+            and self._options.ai
+            and context.activity.type == ActivityTypes.message
+            and context.activity.text
+        ):
+            is_ok = await self._ai.run(context, state)
+            if not is_ok:
+                await self._save_state(context, state)
+                return False
+        return True
+
+    async def _run_after_turn_middleware(self, context: TurnContext, state):
+        for after_turn in self._after_turn:
+            is_ok = await after_turn(context, state)
+            if not is_ok:
+                await self._save_state(context, state)
+                return False
+        return True
+
+    async def _save_state(self, context: TurnContext, state):
```

Comment:
im not sure abstracting this single line function call to another method that calls it adds much value

Reply:
sure, happy to roll that back in. As I mentioned at the top, I went totally verbose on the breakout, not yet knowing the goals or new tests that might be anticipated. Will change shortly.

Reply:
rolled state.save back and tested locally.